src/uct/ib/rdmacm/rdmacm_cm.c

```diff
@@ -65,52 +67,272 @@ ucs_status_t uct_rdmacm_cm_reject(struct rdma_cm_id *id)
+            if (ctx->dummy_qp_ctx.cq == NULL) {
+                ucs_error("ibv_create_cq() failed: %m");
+                status = UCS_ERR_IO_ERROR;
+                goto err_free_ctx;
+            }
+        }
+
+        kh_value(&cm->ctxs, iter) = ctx;
     }
 
-    *cq_p = cq;
+    *ctx_p = ctx;
     return UCS_OK;
+
+err_free_ctx:
+    ucs_free(ctx);
+err_kh_del:
+    kh_del(uct_rdmacm_cm_qp_contexts, &cm->ctxs, iter);
+out:
+    return status;
 }
 
-void uct_rdmacm_cm_cqs_cleanup(uct_rdmacm_cm_t *cm)
+ucs_status_t
+uct_rdmacm_cm_reserved_qpn_blk_alloc(uct_rdmacm_cm_t *cm,
+                                     struct ibv_context *verbs,
+                                     struct uct_rdmacm_cm_qp_context *ctx,
+                                     struct uct_rdmacm_cm_reserved_qpn_blk **blk_p)
 {
-    struct ibv_cq *cq;
-    int ret;
+    ucs_status_t status = UCS_ERR_UNSUPPORTED;
+
+#if HAVE_DECL_MLX5DV_IS_SUPPORTED
+    char in[UCT_IB_MLX5DV_ST_SZ_BYTES(create_reserved_qpn_in)]   = {};
+    char out[UCT_IB_MLX5DV_ST_SZ_BYTES(general_obj_out_cmd_hdr)] = {};
+    struct uct_rdmacm_cm_reserved_qpn_blk *blk;
+    void *attr;
+    uint32_t qpns_per_obj;
+
+    qpns_per_obj = uct_rdmacm_cm_get_qpns_per_obj(ctx->reserved_qpn_ctx);
+    if (qpns_per_obj > UCT_RDMACM_RESERVED_QPN_BLK_BMP_BITS) {
+        ucs_error("reserved qpn bit map length is not enough");
+        return UCS_ERR_INVALID_PARAM;
+    };
+
+    blk = ucs_calloc(1, sizeof(*blk), "reserved_qpn_blk");
+    if (!blk) {
+        return UCS_ERR_NO_MEMORY;
+    }
 
-    kh_foreach_value(&cm->cqs, cq, {
-        ret = ibv_destroy_cq(cq);
-        if (ret != 0) {
-            ucs_warn("ibv_destroy_cq() returned %d: %m", ret);
-        }
-    });
+    UCS_BITMAP_CLEAR(&blk->bmp);
+
+    attr = UCT_IB_MLX5DV_ADDR_OF(create_reserved_qpn_in, in, hdr);
+    UCT_IB_MLX5DV_SET(general_obj_in_cmd_hdr,
+         attr, opcode, UCT_IB_MLX5_CMD_OP_CREATE_GENERAL_OBJECT);
+    UCT_IB_MLX5DV_SET(general_obj_in_cmd_hdr,
+         attr, obj_type, UCT_IB_MLX5_OBJ_TYPE_RESERVED_QPN);
+    UCT_IB_MLX5DV_SET(general_obj_in_cmd_hdr,
+         attr, log_obj_range, ctx->reserved_qpn_ctx.log_reserved_qpn_granularity);
+
+    blk->obj = mlx5dv_devx_obj_create(verbs, in, sizeof(in),
+                                      out, sizeof(out));
+    if (!blk->obj) {
+        ucs_error("mlx5dv_devx_obj_create(CREATE_GENERAL_OBJECT, type=RESERVED_QPN) failed, syndrome %x: %m",
+                  UCT_IB_MLX5DV_GET(create_mkey_out, out, syndrome));
+        status = UCS_ERR_IO_ERROR;
+        goto err_free_blk;
+    }
+
+    blk->first_qpn = UCT_IB_MLX5DV_GET(general_obj_out_cmd_hdr, out, obj_id);
+    blk->next_avail_qpn_offset = 0;
+
+    *blk_p = blk;
+    return UCS_OK;
+
+err_free_blk:
+    ucs_free(blk);
+#endif
+
+    return status;
+}
+
+ucs_status_t
+uct_rdmacm_cm_reserved_qpn_blk_free(struct uct_rdmacm_cm_reserved_qpn_blk *blk)
+{
+#if HAVE_DECL_MLX5DV_IS_SUPPORTED
+    if (mlx5dv_devx_obj_destroy(blk->obj)) {
```

Comment:
assert that the bitmap is all clear before releasing the object

Reply:
@yosefe Remove bitmap. Asseted the refcounter is 0.